contracts/cw4-voting/src/contract.rs

```diff
@@ -0,0 +1,194 @@
+        code_id: msg.cw4_group_code_id,
+        msg: to_binary(&cw4_group::msg::InstantiateMsg {
+            admin: Some(env.contract.address.to_string()),
+            members: msg.initial_members,
+        })?,
+        funds: vec![],
+        label: env.contract.address.to_string(),
+    };
+    let msg = SubMsg::reply_on_success(msg, INSTANTIATE_GROUP_REPLY_ID);
+
+    DAO_ADDRESS.save(deps.storage, &info.sender)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_submessage(msg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::MemberChangedHook { diffs } => {
+            execute_member_changed_hook(deps, env, info, diffs)
+        }
+    }
+}
+
+pub fn execute_member_changed_hook(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    diffs: Vec<cw4::MemberDiff>,
+) -> Result<Response, ContractError> {
+    let group_contract = GROUP_CONTRACT.load(deps.storage)?;
+    if info.sender != group_contract {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let total_weight = TOTAL_WEIGHT.load(deps.storage)?;
+    // As difference can be negative we need to keep track of both
+    // In seperate counters to apply at once and prevent underflow
+    let mut positive_difference: u64 = 0;
+    let mut negative_difference: u64 = 0;
+    for diff in diffs {
+        let user_address = deps.api.addr_validate(&diff.key)?;
+        let weight = diff.new.unwrap_or_default();
+        let old = diff.old.unwrap_or_default();
+        // Do we need to add to positive difference or negative difference
+        if weight > old {
+            positive_difference += weight - old;
+        } else {
+            negative_difference += old - weight;
+        }
+        USER_WEIGHTS.save(
+            deps.storage,
+            &user_address,
+            &Uint128::from(weight),
+            env.block.height,
+        )?;
+    }
+    let new_total_weight = total_weight
+        .checked_add(Uint128::from(positive_difference))
+        .map_err(StdError::overflow)?
+        .checked_sub(Uint128::from(negative_difference))
+        .map_err(StdError::overflow)?;
+    TOTAL_WEIGHT.save(deps.storage, &new_total_weight, env.block.height)?;
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::VotingPowerAtHeight { address, height } => {
+            query_voting_power_at_height(deps, env, address, height)
+        }
+        QueryMsg::TotalPowerAtHeight { height } => query_total_power_at_height(deps, env, height),
+        QueryMsg::Info {} => query_info(deps),
+        QueryMsg::GroupContract {} => to_binary(&GROUP_CONTRACT.load(deps.storage)?),
```

Comment:
Same as the staking contract - how would you feel about adding a query for the DAO addresss?